test/conftest.py

```diff
@@ -1,35 +1,166 @@
 from __future__ import absolute_import
 
-import os
-
+import inspect
 import pytest
+import time
 
-from test.fixtures import KafkaFixture, ZookeeperFixture
+from decorator import decorator, decorate
 
+from kafka import MultiProcessConsumer, SimpleConsumer
+from test.fixtures import KafkaFixture, ZookeeperFixture, get_simple_consumer, get_simple_producer, get_keyed_producer, get_base_producer, get_simple_client
+from test.testutil import kafka_version, version_str_to_list, random_string, wait_for_kafka_client_topic_update, wait_for_simple_client_topic_update
 
 @pytest.fixture(scope="module")
 def version():
-    if 'KAFKA_VERSION' not in os.environ:
-        return ()
-    return tuple(map(int, os.environ['KAFKA_VERSION'].split('.')))
-
+    return kafka_version()
 
 @pytest.fixture(scope="module")
-def zookeeper(version, request):
-    assert version
+def zookeeper():
     zk = ZookeeperFixture.instance()
     yield zk
     zk.close()
 
+_params = {}
+def _set_params(func, *args, **kw):
+    ret = func(*args, **kw)
+    return ret
+
+def set_params(**params):
+    global _params
+
+    def real_set_params(func, *args, **kwargs):
+        _params[func.__name__] = params
+        ret = decorate(func, _set_params)
+        return ret
+
+    return real_set_params
```

Comment:
I don't understand what is the purpose of these setters?

Reply:
When I started coding and testing the fixtures I noticed that the fixtures needed different test methods may need different configurations. For example, some tests need a consumer with `auto_offset_reset=earliest` while others need `auto_offset_reset=latest`.

One way to enable that flexibility would be to let each test to explicitly instantiate the fixtures, without using pytest injection. Another way, would be to provide the pytest fixtures the parameters for each instantiation.

These setters were my way to implement that. Using those, the tests could be coded like this:
```
@set_params(simple_consumer={'auto_offset_reset':'earliest'})
def test_consumer_with_earliest(simple_consumer):
    ...

@set_params(simple_consumer={'auto_offset_reset':'latest'})
def test_consumer_with_latest(simple_consumer):
    ...
```
The parameters set this way are exposed to the fixture method as an injected parameter:
```
@pytest.fixture
def simple_consumer(...some_params..., simple_consumer_params):
    ....
```
Which I though to be simpler and easier to code than handle explicit instantiation and destruction on each test method.

This works only for fixtures with `function` scope.

Thoughts?

Reply:
So I agree with the design goal, but can't you skip the setters in favor of `pytest`'s built-in `@pytest.mark.parametrize()` decorator? 

It provides the flexibility to both dynamically assign the fixture params for a test, and also to run a single test multiple times using different input permutations.

Am I missing some additional functionality that the custom setters provide?

Reply:
That's a good question. At first glance the `@set_params` and `@pytest.mark.parametrize` decorators seem equivalent but they work in very different ways. I had a look at the latter before I implemented the former and realized it didn't suit the purpose I had in mind.

`@pytest.mark.parametrize` is used to inject different values into a test method's parameter. Each value will trigger a new execution of the test. For example, consider the following test method:
```
@pytest.mark.parametrize('auto_offset_reset', ['earliest', 'latest'])
@pytest.mark.parametrize('enable_auto_commit', [True, False])
def test_consumer(kafka_consumer, auto_offset_reset, enable_auto_commit)
    ...
```
This method will run 4 times, each time with the following values for the `(auto_offset_reset, enable_auto_commit)` variables: `('earliest', True)`, `('earliest', False)`, `('latest', True)`, `('latest', False)`.
The `kafka_consumer` fixture, however, **will be instantiated the same way every time**. The parameters above are injected into the test method, but not the fixture itself.

The `@set_params` decorator does not cause multiple test executions. Instead, it just passes parameters for the instantiation of the specified fixture. In the example below, the test is run only once and the parameters specified in the `@set_params` call are injected into the `kafka_consumer` fixture, not the test method:
```
@set_params(kafka_consumer={'auto_offset_reset':'earliest', 'enable_auto_commit':True})
def test_consumer(kafka_consumer)
    ... kafka_consumer is intantiated with the parameters specified above^^
```

Reply:
@jeffwidman You mentioned that `@pytest.mark.parametrize` "provides the flexibility to ... dynamically assign the fixture params for a test".

I can see how that would work if the fixture is explicitly instantiated within the test method. I don't know how the parameters could be passed to the fixture instantiation if `pytest` is injecting it into the test method. Is there a way to do that? I may be missing something here.

Reply:
I just had another look at the pytest documentation. I see that the fixtures can be parametrized through the `@pytest.fixture()` decorator. This still doesn't seem to suit the purpose here, since different tests may need different fixture configurations.

I think there's a way to do what we need using the `pytest_generate_tests` method. I'll have a look into this and will post back here.

Reply:
Re-reading this, I understand the problem now. This makes sense why `@pytest.mark.parametrize()` won't work... I don't know if there's a better way or not. If not, can you add a comment to the code explaining why you had to do this? Just briefly summarize your above comments...  Otherwise I can totally see myself down the road forgetting the rationale here and spending a couple of hours trying to change it before I realize the issue.